### PR TITLE
Ipavlopoulos research context

### PR DESCRIPTION
**Pull request by ipavlopoulos**
Notebook to sample wikipedia comments based on indentation.

### CODE REVIEW

File: research/modeling/bert.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
+import tensorflow as tf
+import pandas as pd
+import tensorflow_hub as hub
+import os
+import re
+import numpy as np
+from bert.tokenization import FullTokenizer
```
Comment (sorensenjs) [inline research/modeling/bert.py]: http://google.github.io/styleguide/pyguide.html#22-imports

import modules not classes

File: research/modeling/bert.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+      data["sentiment"].append(re.match("\d+_(\d+)\.txt", file_path).group(1))
+  return pd.DataFrame.from_dict(data)
+
+# Merge positive and negative examples, add a polarity column and shuffle.
+def load_dataset(directory):
+  pos_df = load_directory_data(os.path.join(directory, "pos"))
+  neg_df = load_directory_data(os.path.join(directory, "neg"))
+  pos_df["polarity"] = 1
+  neg_df["polarity"] = 0
+  return pd.concat([pos_df, neg_df]).sample(frac=1).reset_index(drop=True)
```
Comment (sorensenjs) [inline research/modeling/bert.py]: add comment explaining that sample(frac=1) is a random permutation. should add injection of random_state for determinism during testing.

File: research/modeling/bert.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+# Create datasets (Only take up to max_seq_length words for memory)
+train_text = train_df['sentence'].tolist()
+train_text = [' '.join(t.split()[0:max_seq_length]) for t in train_text]
+train_text = np.array(train_text, dtype=object)[:, np.newaxis]
+train_label = train_df['polarity'].tolist()
+
+test_text = test_df['sentence'].tolist()
+test_text = [' '.join(t.split()[0:max_seq_length]) for t in test_text]
+test_text = np.array(test_text, dtype=object)[:, np.newaxis]
+test_label = test_df['polarity'].tolist()
```
Comment (sorensenjs) [inline research/modeling/bert.py]: DRY principle - extract to method this duplicated 4 lines of code.

File: research/modeling/bert.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+def convert_single_example(tokenizer, example, max_seq_length=256):
+    """Converts a single `InputExample` into a single `InputFeatures`."""
+
+    if isinstance(example, PaddingInputExample):
+        input_ids = [0] * max_seq_length
+        input_mask = [0] * max_seq_length
+        segment_ids = [0] * max_seq_length
+        label = 0
+        return input_ids, input_mask, segment_ids, label
```
Comment (sorensenjs) [inline research/modeling/bert.py]: seems like you could return the same list 3 times?

File: research/modeling/bert.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    if len(tokens_a) > max_seq_length - 2:
+        tokens_a = tokens_a[0 : (max_seq_length - 2)]
+
+    tokens = []
+    segment_ids = []
+    tokens.append("[CLS]")
+    segment_ids.append(0)
+    for token in tokens_a:
+        tokens.append(token)
+        segment_ids.append(0)
```
Comment (sorensenjs) [inline research/modeling/bert.py]: tokens.extend(tokens_a)
segment_ids.extend([0] * len(tokens_a))

File: research/modeling/bert.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    segment_ids.append(0)
+
+    input_ids = tokenizer.convert_tokens_to_ids(tokens)
+
+    # The mask has 1 for real tokens and 0 for padding tokens. Only real
+    # tokens are attended to.
+    input_mask = [1] * len(input_ids)
+
+    # Zero-pad up to the sequence length.
+    while len(input_ids) < max_seq_length:
```
Comment (sorensenjs) [inline research/modeling/bert.py]: use extend here too, it's much faster then .appends in a loop.

File: research/modeling/bert.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    
+    return model
+
+def initialize_vars(sess):
+    sess.run(tf.local_variables_initializer())
+    sess.run(tf.global_variables_initializer())
+    sess.run(tf.tables_initializer())
+    K.set_session(sess)
+
+model = build_model(max_seq_length)
```
Comment (sorensenjs) [inline research/modeling/bert.py]: protect these with a main()

**Comment by ipavlopoulos** [comment]
Thank you Jeff! I meant to push only the two .ipynb files, but my local master was ahead and unintentionally more files were added. I was thinking of using this BERT code a starting point and all your comments were to the point. I have created a new correct pull request (#144).

**Pull request closed without merging**